### PR TITLE
felix_v1: int6+zstd, fp16 embed, MLP 3x, seq4096, sliding window eval

### PR DESCRIPTION
**Pull request by felix-ab**
## Summary

Stacking complementary techniques to maximize parameter efficiency within the 16MB artifact limit.

### int6 quantization + zstd-22
- `QUANT_MAX=31` (6-bit range) stored in int8 containers — same memory layout, ~25% smaller values → better compression
- zlib level-9 replaced with zstd level-22 (~10-15% additional savings)
- Net effect: body weights compress to ~14.1MB instead of ~15.8MB for baseline

### fp16 embedding passthrough
- `tok_emb.weight` skips quantization and is stored as fp16 directly
- Embeddings are the most quantization-sensitive tensor (directly maps to logit quality)
- Cost: ~0.5MB at dim=512; worth it for quality insurance

### MLP 3x (default)
- `MLP_MULT=3` instead of 2
- Space freed by int6+zstd enables wider hidden layers within budget
- Adds ~4M parameters vs baseline at same artifact size

### seq4096 training (default)
- `TRAIN_SEQ_LEN=4096` instead of 1024
- Longer context enables model to learn longer-range dependencies
- Required for sliding window eval to show meaningful gains

### Sliding window eval
- `eval_val_sliding_window()` added: scores tokens with maximum left context
- Stride=512: each token scored with up to `seq_len` context tokens
- `forward_logits()` added to GPT class for logit access without loss reduction
- Runs after quantized roundtrip eval and logs bpb improvement delta
- Free gain at eval time — no artifact cost

## Artifact budget
```
Body: 21.4M params @ int6+zstd → ~14.1MB
Embedding: 512×1024 fp16 → 0.5MB  
Total: ~14.6MB (under 16MB)
```

## Local validation
- Baseline reproduces on Apple Silicon (M4 Pro)
- int6 quant gap confirmed <0.01 bpb on smoke test
- Script is syntax-clean, 1263 lines (under 1500 limit)
- `zstandard` falls back to `zlib` gracefully if not installed

## Status
Draft — awaiting H100 training run to report final val_bpb.

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — felix_v1: int6+zstd, fp16 embed, MLP 3x, seq4096, sliding window eval

**BPB:** (not parsed — see PR title) | **Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

**What I found in the code** (head SHA `dd0269c3951f`, file `records/track_10min_16mb/felix_v1/train_gpt.py`):

Static code review found no TTT adaptation function, no SLOT optimization loop, no n-gram-cache class, and no pre-quant val-token fine-tune. The eval path uses the standard sliding-window stride-64 pattern. The submission is a pure-neural architecture iteration on the standard SP1024/SP4096/SP8192 baseline.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 0.02s, dim=512, layers=9, vocab=1024, code=53262 B, SMOKE_TEST_PASS

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the classification pass — this looks like a clean pure-neural iteration on the standard baseline.

*Auto-classification caveat:* this review was drafted by the AST-based classifier. If there's a non-standard eval mechanism (logit postprocessing, hedge mixing, etc.) that I missed because it's factored into a helper file or a non-standard function name, please flag it and I'll re-run the audit manually.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 0.02s, dim=512, layers=9, vocab=1024, code=53262 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*
